cmd/clusterctl/client/cluster/topology.go

```diff
@@ -0,0 +1,613 @@
+}
+
+// ensure topologyClient implements TopologyClient.
+var _ TopologyClient = &topologyClient{}
+
+// newTopologyClient returns a TopologyClient.
+func newTopologyClient(proxy Proxy, inventoryClient InventoryClient) TopologyClient {
+	return &topologyClient{
+		proxy:           proxy,
+		inventoryClient: inventoryClient,
+	}
+}
+
+// DryRunInput defines the input for DryRun function.
+type DryRunInput struct {
+	Objs              []*unstructured.Unstructured
+	TargetClusterName string
+}
+
+// Modification defined the modification observed on an object.
+type Modification = dryrun.Modification
+
+// Changes defines all the changes detected by the Dryrun execution.
+type Changes = dryrun.Changes
+
+// DryRunOutput defines the output of the DryRun function.
+type DryRunOutput struct {
+	AffectedClusters       []client.ObjectKey
+	AffectedClusterClasses []client.ObjectKey
+	TargetCluster          client.ObjectKey
+	*Changes
+}
+
+func (t *topologyClient) DryRun(in *DryRunInput) (*DryRunOutput, error) {
+	ctx := context.TODO()
+	if err := t.validate(in); err != nil {
+		return nil, errors.Wrap(err, "dry run input failed validation")
+	}
+
+	// If there is a reachable apiserver with CAPI installed fetch a client for the server.
+	// This client will be used as a fall back client when looking for objects that are not
+	// in the input.
+	// Example: This client will be used to fetch the underlying ClusterClass when the input
+	// only has a Cluster object.
+	var c client.Client
+	if err := t.proxy.CheckClusterAvailable(); err == nil {
+		if initialized, err := t.inventoryClient.CheckCAPIInstalled(); err == nil && initialized {
+			c, err = t.proxy.NewClient()
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to create a client to the cluster")
+			}
+		}
+	}
+	if err := t.prepareInput(in, c); err != nil {
+		return nil, errors.Wrap(err, "dry run failed preparing input")
+	}
+	if err := t.defaultAndValidate(in, c); err != nil {
+		return nil, errors.Wrap(err, "dry run failed defaulting and validation on input objects")
+	}
+
+	objs := []client.Object{}
+	for _, o := range in.Objs {
+		objs = append(objs, o)
+	}
+	for _, o := range t.generateCRDs(in) {
+		objs = append(objs, o)
+	}
+	dryRunClient := dryrun.NewClient(c, objs)
+	// Calculate affected ClusterClasses.
+	affectedClusterClasses, err := t.affectedClusterClasses(in, dryRunClient)
+	if err != nil {
+		return nil, errors.Wrap(err, "dry run failed calculating affected ClusterClasses")
+	}
+	// Calculated affected Clusters
+	affectedClusters, err := t.affectedClusters(in, dryRunClient)
+	if err != nil {
+		return nil, errors.Wrap(err, "dry run failed calculating affected Clusters")
+	}
+
+	res := &DryRunOutput{
+		AffectedClusters:       affectedClusters,
+		AffectedClusterClasses: affectedClusterClasses,
+		Changes:                &dryrun.Changes{},
+	}
+
+	// Calculate the target cluster object.
+	// Full changeset is only generated for the target cluster.
```

Comment:
Does this mean we're only getting a full diff if we change an individual cluster but not if we change a clusterclass? If yes, why?

Reply:
Okay, so just to confirm. If I only change the ClusterClass but set the --cluster flag, I will get the full diff for that cluster? Sounds perfect

Reply:
>Okay, so just to confirm. If I only change the ClusterClass (and no cluster) but set the --cluster flag, I will get the full diff for that cluster? Sounds perfect

Yes.